[flink-5568] [Table API & SQL]Introduce interface for catalog, and provide an in-memory implementation. Integrate external catalog with calcite catalog

beyond1920 opened the pull request:
This pr aims to introduce interface for catalog, and provide an in-memory implementation for test and develop, finally integrate external catalog with calcite catalog.
The main change including:
1. Introduce ExternalCatalog abstraction, including introduce ExternalCatalogDatabase as database   in catalog and ExternalCatalogTable as table in catalog.
2. Provide an in-memory implementation for test and develop.
3. Introduce ExternalCatalogSchema which is an implementation of Calcite Schema interface. It registers database in ExternalCatalog as calcite Schemas, and tables in a database as Calcite table.
4. Add ExternalCatalogCompatible annotation. The TableSource with this annotation represents it could be converted to or from externalCatalogTable. ExternalCatalogTableConverter is the converter between externalCatalogTable and tableSource.
5. Introduce CatalogTableHelper utility. It has two responsibilities:  (a)automatically find the TableSources which are with ExternalCatalogCompatible annotation. (b)convert an ExternalCatalogTable instance to a TableSourceTable instance.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalog.scala:
```diff
@@ -0,0 +1,108 @@
+    *
+    * @param table description of table which to alter
+    */
+  def alterTable(table: ExternalCatalogTable): Unit
```
Add an `ignoreIfNotExists` option?

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalog.scala:
```diff
@@ -0,0 +1,108 @@
+  /**
+    * Alters existed database into external Catalog
+    */
+  def alterDatabase(db: ExternalCatalogDatabase): Unit
```
Add an `ignoreIfNotExists` option?

fhueske commented on flink-libraries/flink-table/src/main/java/org/apache/flink/table/annotation/ExternalCatalogCompatible.java:
```diff
@@ -0,0 +1,63 @@
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Public
+public @interface ExternalCatalogCompatible {
```
I think we should add a version field here as well to support multiple versions of Converters.
This might become important when the `TableSource` evolves or if we have different `TableSources` or for the same table type. 

So either 
- `[$tableType:$version -> TableConverter]+`, which would allow support for different table types or 
- `$tableType, [$version -> TableConverter]+`, which limits a `TableSource` to one table type but multiple versions of it.

What do you think @beyond1920, @KurtYoung ?

beyond1920 replied:
could we define tableType is identifier, it includes version information.  For example, kafka0.8/ kafka0.9 / kafka1.0 or hive1.0/ hive2.0.

fhueske replied:
I thought about the version and agree that we do not necessarily need a dedicated version field.
However, I think we need a bit more "safety" in the translation process since the converters will rely on certain properties to be set and the link between catalog table and converter is just a string. IMO, we need some kind of verification that the properties of a `ExternalCatalogTable` are compatible with the requirements of a converter.

What do you think about the following approach:
- We extend the `TableSourceConverter` interface by a few methods to describe its requirements:
```
def requiredProperties: Array[String]
def optionalProperites: Array[String]
def getPropertyDescription(p: String): String
```
- we remove the `ExternalCatalogCompatible` annotation
- we add a `TableType` annotation which carries a String for the table type name and which is annotated to the `TableSourceConverter` (not to the TableSource!). 

By doing this, we scan for converters instead of table sources. IMO, this has the benefit that the required properties of a table type (i.e., the identifier string) are tied to the converter, which knows how to handle the ExternalTable and convert it into a TableSource. The type of the TableSource is not really relevant. It also means that there is one place to go when looking for the requirements of a table type (the converter).

What do you think @beyond1920 and @KurtYoung ?

beyond1920 replied:
This suggestion is pretty good, thanks

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/utils/CommonTestData.scala:
```diff
@@ -68,6 +62,90 @@ object CommonTestData {
   }
 
+  def getMockedFlinkExternalCatalog: ExternalCatalog = {
+    val csvRecord1 = Seq(
```
`csvRecord1` -> `csvTable1`

beyond1920 replied:
val csvRecord1 = Seq() is just csv data, maybe it's better to name it as csvDataRecords?

fhueske replied:
fine with me

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/utils/CommonTestData.scala:
```diff
@@ -68,6 +62,90 @@ object CommonTestData {
+      )
+    )
+
+    val csvRecord2 = Seq(
```
`csvRecord2` -> `csvTable2`

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/utils/CommonTestData.scala:
```diff
@@ -68,6 +62,90 @@ object CommonTestData {
+        "rowDelim" -> "$"
+      )
+    )
+    val catalog = mock(classOf[ExternalCatalog])
```
Couldn't we use the `InMemoryExternalCatalog` here instead of mocking one?

beyond1920 replied:
Good idea

beyond1920 commented on flink-libraries/flink-table/pom.xml:
```diff
@@ -92,7 +92,12 @@ under the License.
 		</dependency>
 
-
+		<dependency>
```
Yes, this problem bothers me, too. the license of reflections jar is  WTFPL, does that means we would do anything, how to check whether it is compatible with AL2? I noticed that the reflections jar is already referenced in  flink-parent.pom, but only for tests, and would not be included in any flink jars. My original thought is not only use this jar, but also include it (like calcit) in the flink-table-{version}.jar.

beyond1920 replied:
fabian,about reflections jar, its license is WTFPL, WTFPL only says 'Everyone is permitted to copy and distribute verbatim or modified
copies of this license document', however it does not say it compatible with ASL explicitly.
Besides, I also want to include the class of this jar into the generated flink-table-{version}.jar, just like calcite class. I don’t know whether it is allowed.

fhueske replied:
Here is what I found out. WTFPL is compatible with AL2 (https://www.apache.org/legal/resolved.html).

Maven Central lists the dependency as BSD licensed (https://search.maven.org/#artifactdetails%7Corg.reflections%7Creflections%7C0.9.10%7Cjar) which is also OK.
However, it seems that some of its transitive dependencies are LGPL licensed which would not work. So maybe we have to exclude those.

Copying a file and relicensing should be OK under WTFPL. However, we should clearly indicate the origin of the code.

beyond1920 replied:
thanks a lot, I would exclude com.google.code.findbugs.* which license is LGPL.

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalogSchema.scala:
```diff
@@ -0,0 +1,166 @@
+
+  override def getTableNames: util.Set[String] = util.Collections.emptySet[String]
+
+  override def contentsHaveChangedSince(lastCheck: Long, now: Long): Boolean = false
```
set contentsHaveChangedSince to false because We want to fetch the latest informations from the underlying ExternalCatalog instead of using calcite caches. Because tables list and databases list of ExternalCatalog may changed anytime.

fhueske replied:
Right, my understanding of the parameter is that Calcite asks "Has the content changed?" and if we say "false" it will look into the cache. If we tell Calcite "true" (the content has changed) it will go against the external catalog. That's why I was thinking that `true` is the right return value. Can you check Calcite's behavior for this function?

beyond1920 replied:
Oh, yes, my bad. Would change it.

beyond1920 commented:
@fhueske, thanks for your review. I changed the pr based on your suggestions, except for one point.
About adding the version field to ExternalCatalogCompatible, could we define tableType is identifier, it includes version information. For example, kafka0.8/ kafka0.9 / kafka1.0 or hive1.0/ hive2.0.

KurtYoung commented:
Hi @fhueske , i like your propose about moving the annotation from `TableSource` to `TableSourceConverter`. Lets do it this way. 
BTW, i noticed that you offered three possible methods to the `TableSourceConverter`, i can only imagine `def requiredProperties: Array[String]
` is necessary for now. It can help validating the converter and to decide which converter we should use when multiple converters have the same `TableType`.

fhueske commented:
Hi @KurtYoung, you are right. Only `requiredProperties()` would be required to verify properties. 

I thought that the other two methods would be a good way to define the parameters of the converter. They could be used to print a usage message or details when the properties are not matched. We can also leave those out if you think that the implementation overhead does not correspond to the gains.

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/CatalogTableHelper.scala:
```diff
@@ -0,0 +1,138 @@
+
+  // config file to specifier the scan package to search tableSources
+  // which is compatible with external catalog.
+  private val tableSourceConfigFileName = "externalCatalogTable.properties"
```
Hi, fabian. I'm a little confused. Which method do you means?
1. we could add a pkgsToScan field to TableConfig, it's response of Users to specify the value of pkgsToScan field. such as tableConfig.setPkgsToScan("org.apache.flink.table.sources","org.apache.flink.streaming.connectors.kafka")?
Then users should know exactly which module every needed converter belongs to. 
2. We could add a converterConfigFileName in TableConfig, it's response of Users to specify the value of pkgsToScan field. such as tableConfig.setConverterConfigFileName("externalCatalogTable.properties")?
The way is a little strange because the file name is fixed based on the convention, may not changed.

fhueske replied:
Sorry, I think I confused something there. 
The approach to have a resource file in each Module / Jar that offers converters is good. Hard-coding the name of the resource file is fine.

Thinking this one step further, would it make sense to not use scanning at all but exactly specify the Converter classes? Then we would not need to use the reflection library to scan the class path. Just as an idea.

beyond1920 replied:
fabian,  register all needed Converter classes instead of use scanning can work, too. However  if somebody adds new tableSourceConverters , such as parquetTableSourceConverter or else, users need to change the code or config file to including the new added Converter?

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalog.scala:
```diff
@@ -0,0 +1,114 @@
+    * Adds table into external Catalog
+    *
+    * @param table          description of table which to create
+    * @param ignoreIfExists whether to ignore operation if table already exists
```
Good advice, I would change the comment.

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalog.scala:
```diff
@@ -0,0 +1,114 @@
+    * @param table          description of table which to create
+    * @param ignoreIfExists whether to ignore operation if table already exists
+    */
+  def createTable(table: ExternalCatalogTable, ignoreIfExists: Boolean): Unit
```
Hi, fabian. Even for the integration with Calcite, getTable(), getDataBase(), and listTables() are sufficient, however an complete ExternalCatalog should be responsible for CRUD operations on db/table. For some readonly ExternalCatalog, we could choose to not supported createX, dropX, andalterX methods , for example, throw UnsupportedOperationException.

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalog.scala:
```diff
@@ -0,0 +1,114 @@
+    * @param tableName table name
+    * @return table
+    */
+  def getTable(dbName: String, tableName: String): ExternalCatalogTable
```
I would add more comments of the method. Thanks!

beyond1920 commented:
@fhueske , thanks for your review. I updated the pr based on your comments.
Your suggestion about moving the annotation from TableSource to TableSourceConverter is good, I changed the pr in this way.  
About not use scanning at all but exactly specify the Converter classes. It can work, too. However, if somebody adds new tableSourceConverters , such as parquetTableSourceConverter or else, users need to change the code or config file to register new added Converters, right? However scanning method can finds all converters automatically.
Let me know what's your opinion, thanks.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSourceConverter.scala:
```diff
@@ -0,0 +1,73 @@
+@TableType(value = "csv")
+class CsvTableSourceConverter extends TableSourceConverter[CsvTableSource] {
+
+  private val required: JSet[String] = ImmutableSet.of("path")
```
I'd add fieldDelim and rowDelim to the required parameters as well.

beyond1920 commented:
@fhueske , thanks for your review. I already updated the pr based on your comments.

fhueske commented:
Thanks for the update @beyond1920.
The PR looks good to me. 
@twalthr do you also want to have a look at this PR?

twalthr commented:
Not to forget: we also need some good documentation for the website.

fhueske commented:
Good point about the documentation. I think this should be added with PR #3409 when registering external catalogs is exposed via the `TableEnvironment`.

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/CrudExternalCatalog.scala:
```diff
@@ -0,0 +1,110 @@
+    */
+  @throws[DatabaseNotExistException]
+  @throws[TableNotExistException]
+  def dropTable(dbName: String, tableName: String, ignoreIfNotExists: Boolean): Unit
```
Take the usage into consideration, when we create database or alter database, we may need to specify other properties except for databaseNames, So the input parameters to createDatabase and alterDatabase is a ExternalCatalogDatabase instance which includes databaseName and other properties.
However, when we drop a database, name is enough.
Take ddl for example,
CREATE DATABASE czech_slovak_names 
  CHARACTER SET = 'keybcs2'
  COLLATE = 'keybcs2_bin';

Drop (DATABASE) [IF EXISTS] database_name;

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalCatalogTable.scala:
```diff
@@ -0,0 +1,76 @@
+case class ExternalCatalogTable(
+    identifier: TableIdentifier,
+    tableType: String,
+    schema: DataSchema,
```
ColumnSchema sounds like a schema to describe a column, not the schema to describe a table. Maybe TableSchema is better, however there is already a class named TableSchema. So I use DataSchema.

fhueske replied:
Actually, I think the available `TableSchema` class can be reused here.
Would that work @beyond1920 ?

beyond1920 replied:
right, thanks for reminder me.

beyond1920 commented on flink-libraries/flink-table/src/main/resources/tableSourceConverter.properties:
```diff
@@ -0,0 +1,19 @@
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
```
Ok.

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/catalog/ExternalTableSourceUtil.scala:
```diff
@@ -0,0 +1,153 @@
+            .getTypesAnnotatedWith(classOf[TableType])
+        clazzWithAnnotations.asScala.foreach(clazz =>
+          if (classOf[TableSourceConverter[_]].isAssignableFrom(clazz)) {
+            if (Modifier.isAbstract(clazz.getModifiers()) ||
```
Yes, I would add check on it based on InstantiationUtil.

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSourceConverter.scala:
```diff
@@ -0,0 +1,73 @@
+    params.get("rowDelim").foreach(csvTableSourceBuilder.lineDelimiter)
+    params.get("quoteCharacter").foreach(quoteStr =>
+      if (quoteStr.length != 1) {
+        throw new IllegalArgumentException("the value of param quoteCharacter is invalid")
```
Ok

beyond1920 commented:
@twalthr , thanks for the review. I have updated the pr based on your suggestion. I would add documents later in the pr #3409. About the name of ExternalCatalog, I notice that there are three kinds of catalog at flink now, the first one is calcite catalog, the second one is function catalog, the third one is the external catalog. I add 'external' prefix to distinguish them.

KurtYoung commented:
Seems good to merge now, merging...
